### PR TITLE
[exporter/signalfx] Translate metrics together instead of individually

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

Translation rules were originally done on metrics individually. This means that only datapoints within the same metric could be operated on together. This is insufficient for the new host metrics format removing the direction dimension. To properly form metrics to match their previous format, translations must be done across metrics, especially aggregations. This only changes the fact that multiple MTS will be operated on at the same time, it doesn't change any translation rules. The solution is to do aggregations at the `ScopeMetric` metric level, so metrics within the same `ScopeMetric` can be aggregated.

This change also properly changes test metric data to be their proper format, moving metrics into corresponding `ResourceMetrics` and `ScopeMetrics`, instead of all being lumped into the same slices.

This is an internal refactoring and testing change that should not impact any input or output data. Because of this, I don't think there should be any CHANGELOG entry.

**Link to tracking Issue:** <Issue number if applicable>

This PR is the first step to solving issue #12641

**Testing:** <Describe what testing was performed and which tests were added.>

Ensured existing tests pass, data being emitted is the same as before change.

**Documentation:** <Describe the documentation added.>

NA since this is a refactor without any external impact.

### CODE REVIEW

**Comment by dmitryax** [comment]
@crobert-1 PTAL at the linter failures

File: exporter/signalfxexporter/internal/translation/converter.go [hunk 1]
```diff
@@ -88,12 +88,12 @@ func (c *MetricsConverter) MetricsToSignalFxV2(md pmetric.Metrics) []*sfxpb.Data
 			ilm := rm.ScopeMetrics().At(j)
 			for k := 0; k < ilm.Metrics().Len(); k++ {
 				dps := c.translator.FromMetric(ilm.Metrics().At(k), extraDimensions)
-				dps = c.translateAndFilter(dps)
 				sfxDataPoints = append(sfxDataPoints, dps...)
 			}
 		}
 	}
 
+	sfxDataPoints = c.translateAndFilter(sfxDataPoints)
```
Comment (dmitryax) [inline exporter/signalfxexporter/internal/translation/converter.go:96]: Given that a lot of unrelated resource metrics can be merged together by the batch processor, I think we should move it to the resource scope, right after `for j := 0; j < rm.ScopeMetrics().Len(); j++ {}`

File: exporter/signalfxexporter/internal/translation/translator.go [hunk 2]
```diff
@@ -682,6 +683,65 @@ func aggregateDatapoints(
...
 	}
 
+	var aggregatedDatapoints []*sfxpb.DataPoint
+
+	// For every data point, need to find every datapoint it matches, so all matches can be aggregated
+	// together. Datapoints can only be used once.
+	indexUsed := make([]bool, len(dps))
+	for i := range indexUsed {
+		indexUsed[i] = false
+	}
```
Comment (dmitryax) [inline exporter/signalfxexporter/internal/translation/translator.go:693]: I don't think this is needed, slice items are guaranteed to be initiated with zero values in go

File: exporter/signalfxexporter/factory_test.go [hunk 3]
```diff
@@ -207,15 +207,15 @@ func TestDefaultTranslationRules(t *testing.T) {
 	require.Len(t, dps, 1)
 	require.Equal(t, 40.0, *dps[0].Value.DoubleValue)
 
-	// system.network.operations.total new metric calculation
+	// system.disk.operations.total new metric calculation
 	dps, ok = metrics["system.disk.operations.total"]
-	require.True(t, ok, "system.network.operations.total metrics not found")
+	require.True(t, ok, "system.disk.operations.total metrics not found")
```
Comment (dmitryax) [inline exporter/signalfxexporter/factory_test.go:212]: Why is this changed?

File: exporter/signalfxexporter/internal/translation/translator.go [hunk 4]
```diff
@@ -671,8 +671,9 @@ func (mp *MetricTranslator) translateDimension(orig string) string {
 	return orig
 }
 
-// aggregateDatapoints aggregates datapoints assuming that they have
-// the same Timestamp, MetricType, Metric and Source fields.
+// aggregateDatapoints makes sure all Metrics with matching Timestamp, MetricType, Metric and Source fields
+// are aggregated.
+// NOTE: Assumes all input datapoints share metric name.
 func aggregateDatapoints(
```
Comment (dmitryax) [inline exporter/signalfxexporter/internal/translation/translator.go:677]: Actually if we move `sfxDataPoints = c.translateAndFilter(sfxDataPoints)` to the scope of one resource as I suggested above, I don't think this pre-aggregation is needed at all. It'll be the same input for `aggregateDatapoints` as before because each `pmetric.ResourceMetrics` object has only one `pmetric.Metric` of a particular name

File: exporter/signalfxexporter/factory_test.go (hunk 3 above)
Comment (dmitryax) [reply exporter/signalfxexporter/factory_test.go:212]: nvm. I see it's just an opportunistic fix

File: exporter/signalfxexporter/internal/translation/converter.go (hunk 1 above)
Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/converter.go:96]: Will do, thanks for pointing this out.

File: exporter/signalfxexporter/internal/translation/translator.go (hunk 2 above)
Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/translator.go:693]: Good to know, I'll remove this.

File: exporter/signalfxexporter/internal/translation/translator.go (hunk 4 above)
Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/translator.go:677]: I did some investigation, and this is what I found. The name check is required in the `case ActionAggregateMetric` to make sure the rule that is being checked actually applies to each metric. However, with the collector running a `hostmetrics` receiver, this check never fails: 
```
// Datapoints match if they have same name, type, time, and source.
if metricTypesMatch &&
    dpsToAggregate[0].Timestamp == dps[j].Timestamp &&
    dpsToAggregate[0].Source == dps[j].Source {
```

So theoretically you're right, we might be able to remove this entire method. I think it's a fair assumption that if metrics have the same name they'll also have the same type. However, I'm more worried about the `timestamp` and `source` attributes.

Is it always guaranteed that datapoints within the same metric will have identical sources and timestamps?

Comment (dmitryax) [reply exporter/signalfxexporter/internal/translation/translator.go:677]: > Is it always guaranteed that datapoints within the same metric will have identical sources and timestamps?

Yes, scraping receivers are always set the same timestamp on one resource metric. I'm not sure what `Source` is. Do we even set this field in signalfx exporter?

Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/translator.go:677]: I'm actually not seeing the `Source` field set in the exporter. I would have expected it to be set when converting from pmetric format to signalfx proto format, but it's never referenced. From this, I think it's safe to remove this whole method. I'll test again as well.

**Comment by crobert-1** [comment]
_Investigating test failures_

File: exporter/signalfxexporter/internal/translation/converter.go [hunk 5]
```diff
@@ -88,13 +88,13 @@ func (c *MetricsConverter) MetricsToSignalFxV2(md pmetric.Metrics) []*sfxpb.Data
 			ilm := rm.ScopeMetrics().At(j)
 			for k := 0; k < ilm.Metrics().Len(); k++ {
 				dps := c.translator.FromMetric(ilm.Metrics().At(k), extraDimensions)
-				dps = c.translateAndFilter(dps)
-				sfxDataPoints = append(sfxDataPoints, dps...)
+				sfxInitialDataPoints = append(sfxInitialDataPoints, dps...)
```
Comment (dmitryax) [inline exporter/signalfxexporter/internal/translation/converter.go:91]: Hm... Looks like we need to empty `sfxInitialDataPoints` prior to entering this loop, otherwise we will be appending the same data multiple times to `sfxModifiedDataPoints` :)

Another suggestion: Since it's mostly only one scope metrics object generated by the scrapers, we can move `translateAndFilter` inside this loop. that should simplify things

Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/converter.go:91]: Good catch on emptying the `sfxInitialDataPoints` slice!

As far as moving the `translateAndFilter` call, having it outside the `scope` loop is essentially the purpose of this PR. It's to be able to properly account for the exporter's translations handling the removal of the `direction` dimension.

The question that I need answered then, is it guaranteed that both "directions" for metrics (e.g. `system.disk.io.read` and `system.disk.io.write`) will be in the **same** `ScopeMetric`? If yes, this PR can essentially be just updating tests to match proper metric format, but if not, we have to have the `translateAndFilter` call outside of the `ScopeMetric` loop.

Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/converter.go:91]: I believe I've answered my own question, but feel free to correct me if I've missed it. I ran the collector locally with the following flag: ```--feature-gates=-receiver.hostmetricsreceiver.emitMetricsWithDirectionAttribute,+receiver.hostmetricsreceiver.emitMetricsWithoutDirectionAttribute```, with the `network` scraper setup.

Also, [referencing the data model documentation](https://opentelemetry.io/docs/reference/specification/metrics/datamodel/#opentelemetry-protocol-data-model), the resource in `ResourceMetrics` is referring to the data source, being my local machine in my testing. `ScopeMetrics` is referring to `The instrumentation Scope (e.g., instrumentation library name, version)`. In my case, the scope name is `otelcol/hostmetricsreceiver/network` and `Version` is `latest`.

I think it's pretty safe to say that other hostmetrics will follow this same format, so metrics with different names, but from the same receiver and resource will still be in the same ScopeMetric. I believe this means we can make this PR simply test data format changes to follow proper metric specifications.

Comment (dmitryax) [reply exporter/signalfxexporter/internal/translation/converter.go:91]: > The question that I need answered then, is it guaranteed that both "directions" for metrics (e.g. system.disk.io.read and system.disk.io.write) will be in the same ScopeMetric?

Yes, all the scraper produce only one `ScopeMetrics` object within `ResourceMetrics`, so we can put it inside the loop iterating over ScopeMetrics with no issues (not inside the inner Metrics loop). It'll help to avoid emptying `sfxInitialDataPoints`.

File: exporter/signalfxexporter/factory_test.go [hunk 6]
```diff
@@ -339,7 +338,8 @@ func testMetricsData() pmetric.Metrics {
 	dp12.SetTimestamp(pcommon.NewTimestampFromTime(time.Unix(1596000000, 0)))
 	dp12.SetIntVal(6e9)
 
-	m2 := ms.AppendEmpty()
+	sm2 := md.ResourceMetrics().At(0).ScopeMetrics().AppendEmpty().Metrics()
+	m2 := sm2.AppendEmpty()
```
Comment (dmitryax) [inline exporter/signalfxexporter/factory_test.go:342]: 👍  the test data doesn't represent reality

**Review by dmitryax — COMMENTED** [review]
Where is the actual change now? :)

**Comment by crobert-1** [comment]
> Where is the actual change now? :)

The change at this point is just updating test data format. It could go with another PR that will update translation rules, but I figure it would be good to have the format change attached to the discussion we've had on this PR.

**Comment by dmitryax** [comment]
but we still apply `translateAndFilter` on data points from one `pdata.Metric`, the idea is to apply it on set of data points from one `pdata.ScopeMetrics` object

**Comment by crobert-1** [comment]
> but we still apply `translateAndFilter` on data points from one `pdata.Metric`, the idea is to apply it on set of data points from one `pdata.ScopeMetrics` object

I don't know why, but I got mixed around with this review. You're right, we want to apply it at the `ScopeMetrics` level. For some reason I got confused and thought the inner-most loop was the scope level, rather than individual metric level.

I'll update the description again and move translateAndFilter call.

File: exporter/signalfxexporter/internal/translation/converter.go (hunk 5 above)
Comment (crobert-1) [reply exporter/signalfxexporter/internal/translation/converter.go]: I've moved the call to be at the ScopeMetrics level.

**Comment by crobert-1** [comment]
I've added a benchmark test that takes hostmetrics and times how long it takes to run `MetricsToSignalFxV2`. The only change to this functionality was the scope of the translation loop. 

Here's the performance comparison before and after this change on the same dataset:
Before:
```
BenchmarkMetricConversion-12       	    2332	    504435 ns/op
BenchmarkMetricConversion-12       	    2166	    558897 ns/op
BenchmarkMetricConversion-12       	    1760	    637230 ns/op
```

After:
```
BenchmarkMetricConversion-12       	    2235	    488777 ns/op
BenchmarkMetricConversion-12       	    2451	    488565 ns/op
BenchmarkMetricConversion-12       	    2424	    514079 ns/op
```
I believe this shows there is a slight performance improvement for the signalfx exporter with the new functionality.

**Comment by dmitryax** [comment]
@crobert-1 thanks for adding the benchmark 👍  Can you please post before and after results?

**Comment by dmitryax** [comment]
Nice! Seems like there is a clear improvement. I believe we can merge it regardless of `duration` attribute's changes

File: exporter/signalfxexporter/testdata/json/hostmetrics.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
...receive"}}],"startTimeUnixNano":"1659645944000000000","timeUnixNano":"1660944008800763000","asInt":"0"},{"attributes":[{"key":"device","value":{"stringValue":"en3"}},{"key":"direction","value":{"stringValue":"transmit"}}],"startTimeUnixNano":"1659645944000000000","timeUnixNano":"1660944008800763000","asInt":"0"},{"attributes":[{"key":"device","value":{"stringValue":"en3"}},{"key":"direction","value":{"stringValue":"receive"}}],"startTimeUnixNano":"1659645944000000000","timeUnixNano":"1660944008800763000","asInt":"0"},{"attributes":[{"key":"device","value":{"stringValue":"en4"}},{"key":"direction","value":{"stringValue":"transmit"}}],"startTimeUnixNano":"1659645944000000000","timeUnixNano":"1660944008800763000","asInt":"0"},{"attributes":[{"key":"device","value":{"stringValue":"en4"}},{"key":"direction","value":{"stringValue":"receive"}}],"startTimeUnixNano":"1659645944000000000","timeUnixNano":"1660944008800763000","asInt":"0"},{"attributes":[{"key":"device","value":{"stringValue":"utun1"}},{"key":"direction","value":{"stringValue":"transmit"}}],"startTimeUnixNano":"1659645944000000000","timeUnixNano":"1660944008800763000","asInt":"0"},{"attributes":[{"key":"device","value":{"strin
```
Comment (dmitryax) [inline exporter/signalfxexporter/testdata/json/hostmetrics.json:20]: shouldn't be only one line? I believe only one is being unmarshaled anyway. Please confirm

Comment (crobert-1) [reply exporter/signalfxexporter/testdata/json/hostmetrics.json:20]: You're right! I removed all of the other lines and the metric data and benchmark test results were the same.

I'll update.

File: exporter/signalfxexporter/factory_test.go [hunk 8]
```diff
@@ -315,9 +315,8 @@ func TestCreateMetricsExporterWithEmptyExcludeMetrics(t *testing.T) {
 
 func testMetricsData() pmetric.Metrics {
```
Comment (dmitryax) [inline exporter/signalfxexporter/factory_test.go:316]: One last thing. Do we really need changes to this test case? It looks like it just breaks the metrics into two ResourceMetrics objects starting from `m4` which is not desired I think

Comment (crobert-1) [reply exporter/signalfxexporter/factory_test.go:316]: The goal was for the test metrics to have the same format that they'd have if they were coming from the `hostmetrics` receiver. I broke them into two `ResourceMetrics` objects to differentiate the time, showing that the two objects were coming one after the other, since you wouldn't get the same metric twice, at two different timestamps, within the same `ScopeMetric` object, or `ResourceMetric`.

I could at least update the variable naming to make it more clear? I think the changes should stay though so at least the test is more future proofed for changes.

Comment (dmitryax) [reply exporter/signalfxexporter/factory_test.go:316]: Ok makes sense now!

**Review by dmitryax — APPROVED** [review]
